### PR TITLE
kem: add hybrid KEM support

### PR DESCRIPTION
**Pull request by bwesterb**
We will use this to add the hybrid KEM X25519+Kyber512 later for KEM TLS.

A hybrid KEM is constructed simply by concatenating the shared secrets, cipher texts, etc, as suggested by https://datatracker.ietf.org/doc/draft-ietf-tls-hybrid-design/ for TLS and approved by https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr2.pdf

The disadvantage of this is that if one of the KEMs is weak, that then a part of the shared secret might be weak as well.  So one has to be careful when using the shared secret: it can't be used as the output of a hash.

To derive key pairs deterministically and to encapsulate deterministically, the seeds are derived using SHAKE256 from a seed that has as size the maximum required for the combined schemes.  This is to prevent a non-uniform seed (like such a shared secret could be) to impact just one of the KEMs.

This also adds one useless hybrid KEM: SIKE+Kyber.  This is just to demonstrate the code and we'll remove it once support for X25519 lands. (#170 )

### CODE REVIEW

**Comment by bwesterb** [comment]
@armfazh @claucece @thomwiggers

I lost write access to the repository now I'm not part of the Cloudflare GitHub group anymore.  So I can't assign reviewers.  Could one of you review this?

File: kem/hybrid/hybrid.go
Comment (thomwiggers) [inline kem/hybrid/hybrid.go:37]: I assume temporary because `x25519` still needs to be specified as a KEM?

File: kem/hybrid/hybrid.go
Comment (thomwiggers) [inline kem/hybrid/hybrid.go:218]: considering you're overruling the seed _anyway_ by deriving it through `Shake`, shouldn't you just hardcode the seed size (in `EncapsulationSeedSize()`) of the hybrid scheme to 32 for Shake256? Why allow smaller ones?

File: kem/hybrid/hybrid.go
Comment (thomwiggers) [inline kem/hybrid/hybrid.go:251]: IDK how Go treats list indexing, but could `ct[:firstSize]` or `ct[firstSize:]` do an index-out-of-bounds?

**Review by thomwiggers — CHANGES_REQUESTED** [review]

File: kem/hybrid/hybrid.go
Comment (bwesterb) [reply kem/hybrid/hybrid.go:37]: Yes.

File: kem/hybrid/hybrid.go
Comment (bwesterb) [reply kem/hybrid/hybrid.go:218]: If you combine two schemes which have 128 bit seeds, then you don't really need a 256 bit seed in total.

File: kem/hybrid/hybrid.go
Comment (bwesterb) [reply kem/hybrid/hybrid.go:251]: At the start of the function we check that the size of `ct` is correct.

Comment (thomwiggers) [reply kem/hybrid/hybrid.go:251]: Right, I missed that :+1:

**Comment by claucece** [comment]
@bwesterb I added the reviewers. Pinging @chris-wood also here ;)

**Comment by claucece** [comment]
LGTM. @bwesterb could you add a test just to show how the functionality can be used? 

@armfazh do you think we can merge x25519 as part of a different PR than #170.. just to see if we can play with it in KEM TLS?

**Review by claucece — CHANGES_REQUESTED** [review]
Add a test for showing how the functionality works.

**Comment by bwesterb** [comment]
> LGTM. @bwesterb could you add a test just to show how the functionality can be used?

It's already in there: the temporary scheme `hybrid.KyberSike()` demonstrates how to use it and as it's added to the list of schemes it's included in the tests run for every scheme.  (See schemes_test.go.)

**Comment by claucece** [comment]
ah perfect @bwesterb

**Review by claucece — APPROVED** [review]

**Comment by armfazh** [comment]
> @bwesterb 
> This is just to demonstrate the code and we'll remove it once support for X25519 lands. (#170 )

> @armfazh do you think we can merge x25519 as part of a different PR than #170.. just to see if we can play with it in KEM TLS?

It happens that the KEMs in hpke (170) are too specfic to the application. So, we need to bring back the previous PR (that I declined) for including kems based on X255/X448. 
Also, these new kems will be specific to the application KEMTLS, so let make sure the documentation states these differences.

File: kem/hybrid/hybrid.go
Comment (armfazh) [inline kem/hybrid/hybrid.go:1]: Is it possible to convert this package in a KEM concatenator, rather than listing specific kems inside of it
otherwise, lets rename package `pqhybrid`

File: kem/schemes/schemes.go
Comment (armfazh) [inline kem/schemes/schemes.go:24]: In a side note, I noticed that the documentation doesn't shows what is registered, only if someone opens the source file.

**Comment by armfazh** [comment]
I also added method Public() in the definition of PrivateKey, (you may want to rebase on top of 170)


```
type PrivateKey interface {
	// Returns the scheme for this private key
	Scheme() Scheme

	encoding.BinaryMarshaler
	Equal(PrivateKey) bool
	Public() PublicKey
}
```

File: kem/hybrid/hybrid.go
Comment (bwesterb) [reply kem/hybrid/hybrid.go:1]: I don't think the user should be able to concatenate any KEMs themselves.  This package will list some of them.  I don't see why we have to reflect the fact that they're all PQ in the name.  And why is `pqhybrid` better than `classichybrid`? `classicpqhybrid` is too long.  Just `hybrid` is clear enough I think.

File: kem/schemes/schemes.go
Comment (bwesterb) [reply kem/schemes/schemes.go:24]: I don't understand what you mean.

**Comment by bwesterb** [comment]
> I also added method Public() in the definition of PrivateKey, (you may want to rebase on top of 170)
> 
> ```go
> type PrivateKey interface {
> 	// Returns the scheme for this private key
> 	Scheme() Scheme
> 
> 	encoding.BinaryMarshaler
> 	Equal(PrivateKey) bool
> 	Public() PublicKey
> }
> ```

I'd rather make the changes after #170 is merged (and so doesn't change anymore) or have this one merged first.

File: kem/schemes/schemes.go
Comment (armfazh) [reply kem/schemes/schemes.go:24]: run godoc command and see the documentation of scheme package.

Comment (bwesterb) [reply kem/schemes/schemes.go:24]: It's because `allSchemes` is private.  We could add a testable example for `All()` which prints the names of all schemes.  That would show up in the online documentation (but `go doc` doesn't seem to list examples.)

Comment (armfazh) [reply kem/schemes/schemes.go:24]: I suggest to add a list in the documentation showing schemes. I can include this in the other PR.

go doc  != godoc

Comment (bwesterb) [reply kem/schemes/schemes.go:24]: Then a testable example which outputs the list would be better because that'll force it to keep in synch, whereas a simple list will get out of date.

**Comment by bwesterb** [comment]
@armfazh Updated to API changes.  Ready to merge.

**Review by armfazh — APPROVED** [review]

**Pull request merged**
